src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Sse2.cs

```diff
@@ -22,67 +22,76 @@ public new abstract class X64 : Sse.X64
         {
             internal X64() { }
 
-            public new static bool IsSupported { get => IsSupported; }
+            public new static bool IsSupported { [Intrinsic]get => IsSupported; }
 
             /// <summary>
             /// __int64 _mm_cvtsd_si64 (__m128d a)
             ///   CVTSD2SI r64, xmm/m64
             /// This intrinisc is only available on 64-bit processes
             /// </summary>
+            [Intrinsic]
             public static long ConvertToInt64(Vector128<double> value) => ConvertToInt64(value);
             /// <summary>
             /// __int64 _mm_cvtsi128_si64 (__m128i a)
             ///   MOVQ reg/m64, xmm
             /// This intrinisc is only available on 64-bit processes
             /// </summary>
+            [Intrinsic]
             public static long ConvertToInt64(Vector128<long> value) => ConvertToInt64(value);
```

Comment:
This one shouldn't be x64 specific.

We can still emit the `MOVQ` standalone instruction (which requires the result be stored in memory):
![image](https://user-images.githubusercontent.com/10487869/49192601-8b8c2f00-f330-11e8-881e-615d40806ca7.png)

This is different from the `movd` instruction which has an `movq` overload that is 64-bit only:
![image](https://user-images.githubusercontent.com/10487869/49192623-a2cb1c80-f330-11e8-91da-6d0fc8af8c20.png)

Reply:
Ah, this is comment mistake. This intrinsic generates `cvtss2si`.

Reply:
Why would this intrinsic generate `cvtss2si` (which takes a scalar `float` and returns an `integer`)?

The intrinsic is `_mm_cvtsi128_si64` which takes a scalar `integer` and returns an `integer`, which generates the `movq` instruction as per: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cvtsi128_si64&expand=1851

Reply:
Oh, my mistake, I are talking about `ConvertToInt64(Vector128<long> value)`, not `ConvertToInt64(Vector128<double> value)`. 

If we match C++, this should be a 64-bit only intrinsic, as C++ generate `movq r64, xmm`.

Reply:
I think this is a case where it would be fine, on x86 machines, to just force the store to memory. On x64 machines, we can do whichever is more appropriate (store to memory or to register).

We are already needing to provide the `ToScalar` functionality on x86 anyways (via the helper intrinsics) so it would just make that simpler overall.

Reply:
> We are already needing to provide the ToScalar functionality on x86 anyways (via the helper intrinsics) so it would just make that simpler overall.

Shall we remove these two APIs and only expose `ToScalar`, you did that for `ConvertToSingle`.

Reply:
The difference is that `ConvertToSingle` was exclusively a helper intrinsic (`float` is always in an XMM register and can be folded away completely) while `ConvertToInt64` requires actual codegen (you have to either go from `XMM` to `r64` or to `mem64`)

Reply:
> to just force the store to memory. 

I think this intrinsic should be 64-bit specific (same as C++). This is not only generating an instruction to store to memory but also related to `Int64` representation on 32-bit platforms. Using `Int64` on 32-bit platforms is not perf friendly, so it is not necessary to complicate the intrinsic system.

Reply:
It would be great if @CarolEidt can give some suggestions.

Reply:
I don't have a strong opinion about this, but if pressed I would say that we should just support this on both systems. While we're trying to stay as true as possible to the actual instructions, the distinction between whether operands or results are in memory or in registers is not something that is under the control of the developer. So exposing intrinsics that may generate a store vs. a register definition seems like the reasonable thing to do.

Reply:
> So exposing intrinsics that may generate a store vs. a register definition seems like the reasonable thing to do.

I don't think it's that reasonable. Basically the proposal is to attempt to implement this intrinsic by going through memory. That may sound trivial but it's likely not and I don't think there's any other intrinsic that does something like this. 

What memory? If the user did not provide a memory location so the JIT needs to pull it out from thin air. Basically it has to spill the result to a lclvar, ensure that the lclvar is not enregistered and that the resulting tree is not transformed in some weird way that would separate the lclvar from the intrinsic. And then we need to decompose it somehow. Or perhaps an alternative solution is to use a spill temp.

Either way it sounds like a lot of complication for no benefit.

Reply:
> Either way it sounds like a lot of complication for no benefit.

I agree with @mikedn, users can use `ToScalar` for the functionality on both platforms and use `ConvertToInt64` when they actually want a `movq r64, xmm`.

Reply:
> users can use ToScalar for the functionality on both platforms and use ConvertToInt64 when they actually want a movq r64, xmm.

Or `GetElement(0)`. Looks like we already have multiple ways of doing the same thing.

Reply:
Yes, exactly. `ToScalar` and `GetElement` is implement by `Unsafe` on unsupported platforms, so we should optimize `Unsafe` CQ rather than complicate the whole JIT again for `ConvertToInt64`.

Reply:
As I say, I don't have a strong opinion, but I'm confused about the complexity. As @tannergooding points out, we already have the `ToScalar` helper function. What am I missing that would cause this to be more complex?
And in any case, if it is equivalent to `ToScalar` then do we need this intrinsic at all?

Reply:
> What am I missing that would cause this to be more complex?

As far as I can tell `ToScalar` has "software" fallback. `ConvertToInt64` is supposed to be an intrinsic so it should not have fallback.

> And in any case, if it is equivalent to ToScalar then do we need this intrinsic at all?

Looks to me than all this intrinsic affair morphed from exposing the relevant instructions as they are, without "added value" to a mish-mash of actual intrinsic and various helpers with software fallback. Whatever. But if this results in unnecessary complexity in the JIT then I'd say that it's a bad affair.

Reply:
> we already have the ToScalar helper function. What am I missing that would cause this to be more complex?

`ToScalar` is a platform agnostic helper that does not guarantee the optimal codegen now. @tannergooding is suggesting generating `movq m64, xmm` for `ConvertToInt64` on 32-platform.
Of course, we can implement `ToScalar` via `ConvertToInt64` but not inverse.